drivers/sensor/adi/adxl34x/adxl34x_decoder.c

```diff
@@ -0,0 +1,302 @@
+	default:
+		return -ENOTSUP;
+	}
+	return 0;
+}
+
+/**
+ * Get the size required to decode a given channel
+ *
+ * @param[in] channel The channel to query
+ * @param[out] base_size The size of decoding the first frame
+ * @param[out] frame_size The additional size of every additional frame
+ * @return 0 if successful, negative errno code if failure.
+ * @see sensor_get_decoder(), sensor_decode(), get_size_info()
+ */
+static int adxl34x_decoder_get_size_info(struct sensor_chan_spec channel, size_t *base_size,
+					 size_t *frame_size)
+{
+	switch (channel.chan_type) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+#if defined(CONFIG_ADXL34X_DATA_TYPE_Q31)
+		*base_size = sizeof(struct sensor_three_axis_data);
+		*frame_size = sizeof(struct sensor_three_axis_sample_data);
+#elif defined(CONFIG_ADXL34X_DATA_TYPE_SENSOR_VALUE)
+		*base_size = sizeof(struct sensor_value[3]);
+		*frame_size = sizeof(struct sensor_value[3]);
+#elif defined(CONFIG_ADXL34X_DATA_TYPE_DOUBLE)
+		*base_size = sizeof(struct adxl343_sensor_data);
+		*frame_size = sizeof(struct adxl343_sample_value);
+#endif
+		break;
+	default:
+		return -ENOTSUP;
+	}
+	return 0;
+}
+
+/**
+ * Decode up to @p max_count samples from the buffer
+ *
+ * @param[in] buffer The buffer provided on the @ref rtio context
+ * @param[in] channel The channel to decode
+ * @param[out] fit The current frame iterator
+ * @param[in] max_count The maximum number of channels to decode.
+ * @param[out] data_out The decoded data
+ * @return 0 if successful, negative errno code if failure.
+ * @see sensor_get_decoder(), sensor_decode(), get_frame_count(), get_size_info()
+ */
+static int adxl34x_decoder_decode(const uint8_t *buffer, struct sensor_chan_spec channel,
+				  uint32_t *fit, uint16_t max_count, void *data_out)
+{
+	ARG_UNUSED(fit);
+	ARG_UNUSED(max_count);
+	const struct adxl34x_encoded_data *edata = (const struct adxl34x_encoded_data *)buffer;
+	const struct adxl34x_decoder_header *header = &edata->header;
+	int nr_of_decoded_samples = 0;
+
+	if (channel.chan_idx != 0) {
+		return -ENOTSUP;
+	}
+	if (header->entries == 0) {
+		return -ENOTSUP;
+	}
+	if (channel.chan_type != SENSOR_CHAN_ACCEL_X && channel.chan_type != SENSOR_CHAN_ACCEL_Y &&
+	    channel.chan_type != SENSOR_CHAN_ACCEL_Z &&
+	    channel.chan_type != SENSOR_CHAN_ACCEL_XYZ) {
+		return -ENOTSUP;
+	}
+
+	const uint16_t range_scale = adxl34x_range_conv[header->range];
+	uint8_t offset = 0;
+
+#if defined(CONFIG_ADXL34X_DATA_TYPE_Q31)
+	const uint8_t shift = adxl34x_shift_conv[header->range];
+	struct sensor_three_axis_data *out = (struct sensor_three_axis_data *)data_out;
+
+	out->header.base_timestamp_ns = edata->header.timestamp;
+	out->header.reading_count = header->entries;
+	out->shift = shift;
+
+#elif defined(CONFIG_ADXL34X_DATA_TYPE_SENSOR_VALUE)
+	struct sensor_value *out = (struct sensor_value *)data_out;
+
+#elif defined(CONFIG_ADXL34X_DATA_TYPE_DOUBLE)
```

Comment:
Doubles aren't supported from the driver itself through this API, its always possible to convert a q31 to a double.

Reply:
The Zephyr documentation indicates the q31 notation is experimental, so I wasn't sure what the best approach would be. Therefore I made the output configurable compile time. I have no problem removing this 'feature' and completely switch to q31 notation for the decoder. Shall I completely remove the 'double' and 'sensor_value' support?